### PR TITLE
WIP - Spike integration of Opentracing standard

### PR DESCRIPTION
**Pull request by wadetandy**
This is a _very_ rough pass at implementing the `opentracing` gem, which
is created to follow the [Opentracing Standard](https://opentracing.io/). Tools integrating with opentracing
are supported out of the box by a number of APM tools available in the
industry, and connecting support for others becomes much easier through
the standard interface provided.  This will allow more granular
performance optimization than the log debugger we have in place now.

As I said above, this is still pretty rough and doesn't have specs or
anything, but I wanted to get something out for discussion purposes. I'd
ideally like to use ActiveSupport::Notifications the same way we are
already doign for the debug log, but as far as I know it only supports
point-in-time notifications, as opposed to wrapping the entire timeframe
for a call, which is important for tracing.

Right now the debugger and the tracer are both using slightly different
event names as well (and the things they are tracking don't 100%
overlap yet). Ideally both of these could be resolved before this makes
it into master.

### CODE REVIEW

**Comment by wadetandy** [comment]
@richmolj would love to get some initial thoughts about this. Especially interested in unifying the solution with the log debugger as much as possible.

**Comment by richmolj** [comment]
Looks great! I'm mostly thinking of how we can keep the existing `ActiveSupport::Notifications`, since that way this will not affect the debugger (which has edge cases and no real tests at this point).

> point-in-time notifications, as opposed to wrapping the entire timeframe

It gives you a start and stop time, is that not enough?

**Comment by wadetandy** [comment]
> It gives you a start and stop time, is that not enough?

It's enough _information_, but the way opentracing (and tracing in general) works needs this concept of nested "spans" which can't be calculated after the fact. Or at least can't be without throwing out a ton of the benefit of the toolkit and adding too much re-implementation.

I'm also fine punting on a combined solution for the time being and just saying tracing its its own thing parallel to AR:Notifications & the debugger until we understand the problem a little bit better.

**Comment by richmolj** [comment]
I know [opentracing-rails uses ActiveRecord subscriptions](https://github.com/xeviknal/opentracing-rails/blob/master/lib/opentracing_rails/subscribers/active_record.rb). I can imagine how this might provide less benefit though.

>  tracing its its own thing parallel to AR:Notifications & the debugger

My main thought is tracing is implemented with Notifications or implemented completely independently. Or, maybe trace and broadcast the Notification in the trace code. Point being that the Notifications are a good thing that can be used by anyone for purposes other than tracing.

**Comment by wadetandy** [comment]
> Point being that the Notifications are a good thing that can be used by anyone for purposes other than tracing.

No argument there.  How about something akin to:

```ruby
Graphiti.tracer.trace('foo') do |trace_scope|
  # do stuff
end

class Graphiti::Tracer
  def trace(name)
    full_name = "graphiti.#{name}"

    Opentracing.start_active_span(full_name) do |scope|
      ActiveSupport::Notifications.instrument(full_name) do 
        retval = yield scope
      end
    end

    retval
  end
end
```

Also, looked into the AR tracing code you linked above and the only reason that works as written is because there can never be a nested trace.  Doing this that way would mean we'd never be able to capture the nesting of later things like activerecord & sql queries.

**Comment by wadetandy** [comment]
@richmolj Updated to integrate `Graphti.broadcast` directly into the tracing pipeline so they can be used together. Still needs specs but want to verify we're in agreement on the approach before I finish that off.

File: graphiti.gemspec [hunk 1]
```diff
@@ -24,6 +24,7 @@ Gem::Specification.new do |spec|
   spec.add_dependency 'graphiti_errors', '~> 1.0.beta.1'
   spec.add_dependency 'concurrent-ruby', '~> 1.0'
   spec.add_dependency 'activesupport', ['>= 4.1', '< 6']
+  spec.add_dependency 'opentracing', '~> 0.4.3'
```
Comment (richmolj) [inline graphiti.gemspec:27]: How feasible would it be to make this a separate gem, so people don't have to pull in this dep if they don't want?

**Review by richmolj — COMMENTED** [review]
@wadetandy this seems reasonable as long as we do a bunch of validation everything still works. I'm wondering if we can make this an optional gem, though (maybe this PR becomes defining a pluggable interface, and the gem can plug in)

**Comment by richmolj** [comment]
@wadetandy if there's anything left to do, let's recreate the PR in current Graphiti repo
